File: coding-interview-patterns/scala/src/main/scala/patterns/Greedy.scala
package patterns
object Greedy {
  def mergeIntervals(intervals:Array[Array[Int]]): Array[Array[Int]] = {
    val arr = intervals.sortBy(_(0))
    val out = scala.collection.mutable.ArrayBuffer[Array[Int]]()
    for (iv <- arr){
      if (out.isEmpty || out.last(1) < iv(0)) out += Array(iv(0), iv(1))
      else out.last(1) = out.last(1).max(iv(1))
    }
    out.toArray
  }
  def eraseOverlapIntervals(intervals:Array[Array[Int]]): Int = {
    if (intervals.isEmpty) return 0
    val arr = intervals.sortBy(_(1))
    var end=arr(0)(1); var keep=1
    for (i<-1 until arr.length){ val s=arr(i)(0); val e=arr(i)(1); if (s>=end){ keep+=1; end=e } }
    arr.length - keep
  }
  def minArrows(points:Array[Array[Int]]): Int = {
    if (points.isEmpty) return 0
    val arr = points.sortBy(_(1))
    var end=arr(0)(1); var arrows=1
    for (i<-1 until arr.length){ val s=arr(i)(0); val e=arr(i)(1); if (s>end){ arrows+=1; end=e } }
    arrows
  }
}

File: coding-interview-patterns/python/patterns/greedy.py
def merge_intervals(intervals):
    intervals.sort(); out=[]
    for s,e in intervals:
        if not out or out[-1][1]<s: out.append([s,e])
        else: out[-1][1]=max(out[-1][1],e)
    return out

def erase_overlap_intervals(intervals):
    if not intervals: return 0
    intervals.sort(key=lambda x:x[1])
    end=intervals[0][1]; keep=1
    for s,e in intervals[1:]:
        if s>=end: keep+=1; end=e
    return len(intervals)-keep

def find_min_arrow_shots(points):
    if not points: return 0
    points.sort(key=lambda x:x[1])
    end=points[0][1]; arrows=1
    for s,e in points[1:]:
        if s>end: arrows+=1; end=e
    return arrows
